Fundamentals restructuring (#10929)
aspnet/Docs: aspnetcore/fundamentals/index.md

You can also use the host to run background tasks. For more information, see [Background tasks](xref:fundamentals/host/hosted-services).

## Servers

An ASP.NET Core app uses an HTTP server implementation to listen for HTTP requests. The server surfaces requests to the app as a set of [request features](xref:fundamentals/request-features) composed into an `HttpContext`.

::: moniker range=">= aspnetcore-2.2"

# [Windows](#tab/windows)

ASP.NET Core provides the following server implementations for Windows:

* *Kestrel* is a cross-platform web server. Kestrel is often run in a reverse proxy configuration using [IIS](https://www.iis.net/). In ASP.NET Core 2.0 or later, Kestrel can be run as a public-facing edge server exposed directly to the Internet.
* *IIS HTTP Server* is a server for windows that uses IIS. With this server, the ASP.NET Core app and IIS run in the same process.
* *HTTP.sys* is a server for Windows that isn't used with IIS.

# [macOS](#tab/macos)

ASP.NET Core provides the *Kestrel* cross-platform server implementation. In ASP.NET Core 2.0 or later, Kestrel can be run as a public-facing edge server exposed directly to the Internet.

# [Linux](#tab/linux)

> ASP.NET Core provides the *Kestrel* cross-platform server implementation. In ASP.NET Core 2.0 or later, Kestrel can be run as a public-facing edge server exposed directly to the Internet. Kestrel is often run in a reverse proxy configuration with [Nginx](http://nginx.org) or [Apache](https://httpd.apache.org/).

Author: Added the sentence to the macOS tab, but these being OS tabs they have special functionality to be selected automatically depending on the user's OS.  So far as I can see, a combined Linux/Mac tab isn't supported.

ASP.NET Core provides the following server implementations:

* *Kestrel* is a cross-platform web server. Kestrel is often run in a reverse proxy configuration using [IIS](https://www.iis.net/). In ASP.NET Core 2.0 or later, Kestrel can be run as a public-facing edge server exposed directly to the Internet.
* *HTTP.sys* is a server for Windows that isn't used with IIS.

# [macOS](#tab/macos)

ASP.NET Core provides the *Kestrel* cross-platform server implementation. In ASP.NET Core 2.0 or later, Kestrel can be run as a public-facing edge server exposed directly to the Internet.

# [Linux](#tab/linux)

ASP.NET Core provides the *Kestrel* cross-platform server implementation. In ASP.NET Core 2.0 or later, Kestrel can be run as a public-facing edge server exposed directly to the Internet. Kestrel is often run in a reverse proxy configuration with [Nginx](http://nginx.org) or [Apache](https://httpd.apache.org/).

---

::: moniker-end

For more information, see [Servers](xref:fundamentals/servers/index).

## Configuration

> ASP.NET Core provides a configuration framework that isn't based on [System.Configuration](xref:System.Configuration) or *web.config*. The framework gets settings as name-value pairs from an ordered set of configuration providers. There are built-in configuration providers for a variety of sources, such as *.json* files, *.xml* files, environment variables, and command-line arguments. You can also write custom configuration providers.

Reviewer: If coming to ASP.NET Core without ASP.NET 4.x experience, this first sentence will be confusing.

For example, you could specify that configuration comes from *appsettings.json* and environment variables. Then when the value of *ConnectionString* is requested, the framework looks first in the *appsettings.json* file. If the value is found there but also in an environment variable, the value from the environment variable would take precedence.

> For managing confidential configuration data such as passwords, ASP.NET Core provides a [Secret Manager tool](xref:security/app-secrets). For production secrets, we recommend [Azure Key Vault](https://docs.microsoft.com/en-us/aspnet/core/security/key-vault-configuration).

Reviewer: ```suggestion
For managing confidential configuration data such as passwords, ASP.NET Core provides a [Secret Manager tool](xref:security/app-secrets). For production secrets, we recommend [Azure Key Vault](/aspnet/core/security/key-vault-configuration).
```

For more information, see [Configuration](xref:fundamentals/configuration/index).

## Options

> Wherever possible, ASP.NET Core follows the *options pattern* for storing and retrieving configuration values. The options pattern uses classes to represent groups of related settings.

Reviewer: ```suggestion
Where possible, ASP.NET Core follows the *options pattern* for storing and retrieving configuration values. The options pattern uses classes to represent groups of related settings.
```

For example, the following code sets WebSockets options:

```csharp

> var options = new WebSocketOptions()  

Reviewer: ```suggestion
var options = new WebSocketOptions  
```

{  
   KeepAliveInterval = TimeSpan.FromSeconds(120),  
   ReceiveBufferSize = 4096
};  
app.UseWebSockets(options);
```

For more information, see [Options](xref:fundamentals/configuration/options).

## Environments

> Execution environments, such as *Development*, *Staging*, and *Production*, are a first-class notion in ASP.NET Core. You can specify the environment an app is running in by setting the ASPNETCORE_ENVIRONMENT environment variable. ASP.NET Core reads that environment variable at app startup and stores the value in an `IHostingEnvironment` implementation. The environment object is available anywhere in the app via DI.

Reviewer: ```suggestion
Execution environments, such as *Development*, *Staging*, and *Production*, are a first-class notion in ASP.NET Core. You can specify the environment an app is running in by setting the `ASPNETCORE_ENVIRONMENT` environment variable. ASP.NET Core reads that environment variable at app startup and stores the value in an `IHostingEnvironment` implementation. The environment object is available anywhere in the app via DI.
```

::: moniker range=">= aspnetcore-2.0"

> Here's sample code from the `Startup` class that configures the app to provide detailed error information only when it runs in development:

Reviewer: ```suggestion
The following sample code from the `Startup` class configures the app to provide detailed error information only when it runs in development:
```

::: moniker-end

The `ILogger` interface lets you pass any number of fields to the logging provider. The fields are commonly used to construct a message string, but the provider can also send them as separate fields to a data store. This feature makes it possible for logging providers to implement [semantic logging, also known as structured logging](https://softwareengineering.stackexchange.com/questions/312197/benefits-of-structured-logging-vs-basic-logging).

For more information, see [Logging](xref:fundamentals/logging/index).

## Routing

A *route* is a URL pattern that is mapped to a handler. The handler is typically a Razor page, an action method in an MVC controller, or a middleware. ASP.NET Core routing gives you control over the URLs used by your app.

For more information, see [Routing](xref:fundamentals/routing).

## Error handling

ASP.NET Core has built-in features for handling errors, such as:

* A developer exception page
* Custom error pages
* Static status code pages
* Startup exception handling

For more information, see [Error handling](xref:fundamentals/error-handling).

::: moniker range=">= aspnetcore-2.1"

## Make HTTP requests

> An implementation of `IHttpClientFactory` is available for creating  `HttpClient` instances. The factory:

Reviewer: ```suggestion
An implementation of `IHttpClientFactory` is available for creating `HttpClient` instances. The factory:
```

* Provides a central location for naming and configuring logical `HttpClient` instances. For example, a *github* client can be registered and configured to access GitHub. A default client can be registered for other purposes.
* Supports registration and chaining of multiple delegating handlers to build an outgoing request middleware pipeline. This pattern is similar to the inbound middleware pipeline in ASP.NET Core. The pattern provides a mechanism to manage cross-cutting concerns around HTTP requests, including caching, error handling, serialization, and logging.
* Integrates with *Polly*, a popular third-party library for transient fault handling.
* Manages the pooling and lifetime of underlying `HttpClientMessageHandler` instances to avoid common DNS problems that occur when manually managing `HttpClient` lifetimes.
* Adds a configurable logging experience (via *ILogger*) for all requests sent through clients created by the factory.

For more information, see [Make HTTP requests](xref:fundamentals/http-requests).

::: moniker-end

## Content root

> The content root is the base path to any private content used by the app, such as its Razor pages and MVC views. By default, the content root is the base path for the executable hosting the app. An alternative location can be specified when [building the host](#host).

Reviewer: ```suggestion
The content root is the base path to any private content used by the app, such as its Razor files. By default, the content root is the base path for the executable hosting the app. An alternative location can be specified when [building the host](#host).
```

::: moniker range="<= aspnetcore-2.2"

For more information, see [Content root](xref:fundamentals/host/web-host#content-root).

::: moniker-end

::: moniker range="> aspnetcore-2.2"

For more information, see [Content root](xref:fundamentals/host/generic-host#content-root).

::: moniker-end

## Web root

The web root (also known as *webroot*) is the base path to public, static resources, such as CSS, JavaScript, and image files. The static files middleware will only serve files from the web root directory (and sub-directories) by default. The web root path defaults to *\<content root>/wwwroot*, but a different location can be specified when [building the host](#host).

> In Razor (*.cshtml*) files, the tilde-slash  `~/` points to the web root. Paths beginning with `~/` are referred to as virtual paths.

Reviewer: ```suggestion
In Razor (*.cshtml*) files, the tilde-slash `~/` points to the web root. Paths beginning with `~/` are referred to as virtual paths.
```
